<a href="https://colab.research.google.com/github/mikeninerbravog/POC_ML_AI-in-Finance/blob/master/POC_ML_FraudDetection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
"""
O código inclui três demonstrações didáticas:
1. **Detecção de Fraudes**: Uso do algoritmo *Isolation Forest* para identificar transações suspeitas.
2. **Previsão de Tendência de Mercado**: Modelo de *Regressão Linear* para prever preços futuros.
3. **Chatbot Financeiro**: Simulação de um assistente virtual que responde perguntas básicas sobre finanças.
"""

import numpy as np
import pandas as pd
from sklearn.ensemble import IsolationForest
from sklearn.linear_model import LinearRegression
import random

# 1. Simulação de Detecção de Fraudes e Transações Válidas
def detect_transactions():
    np.random.seed(42)
    transactions = np.random.normal(100, 20, 100).tolist()  # Transações normais
    transactions.extend([500, 600, 750, 1000])  # Adicionando fraudes
    df = pd.DataFrame({'transaction_value': transactions})

    model = IsolationForest(contamination=0.05)
    df['fraudulent'] = model.fit_predict(df[['transaction_value']])

    frauds = df[df['fraudulent'] == -1]
    valid_transactions = df[df['fraudulent'] == 1]

    print("Transações fraudulentas detectadas:\n", frauds)
    print("Transações válidas detectadas:\n", valid_transactions)
    return df

# 2. Previsão de Tendência de Mercado
def market_trend_prediction():
    np.random.seed(42)
    days = np.arange(1, 101)
    prices = 50 + np.cumsum(np.random.normal(0, 1, 100))  # Simulação de preços

    model = LinearRegression()
    model.fit(days.reshape(-1, 1), prices)
    future_prices = model.predict([[101], [102], [103]])

    print("Previsão para os próximos 3 dias:", future_prices)
    return prices, future_prices

# 3. Chatbot Financeiro Simples
def chatbot_finance(question):
    """
    Chatbot simples para responder perguntas sobre finanças.
    Esse chatbot pode ser expandido para um assistente inteligente e autônomo,
    capaz de fornecer respostas mais sofisticadas e interagir dinamicamente com os usuários.
    O resultado aqui é apenas um exemplo para a PoC.
    """
    responses = {
        "O que é um investimento?": "Investimento é quando você usa dinheiro para comprar algo esperando que ele valha mais no futuro.",
        "Como evitar fraudes financeiras?": "Use cartões seguros, desconfie de ofertas muito boas e monitore suas transações.",
        "O que é um mercado de ações?": "É onde pessoas compram e vendem partes de empresas chamadas ações.",
    }
    return responses.get(question, "Desculpe, não entendi sua pergunta.")

# Executando as funções
detect_transactions()
market_trend_prediction()

# Exemplo de interação com o chatbot
pergunta = "O que é um investimento?"
print(f"Usuário: {pergunta}")
print(f"Chatbot: {chatbot_finance(pergunta)}")


Transações fraudulentas detectadas:
      transaction_value  fraudulent
31          137.045564          -1
74           47.605098          -1
100         500.000000          -1
101         600.000000          -1
102         750.000000          -1
103        1000.000000          -1
Transações válidas detectadas:
     transaction_value  fraudulent
0          109.934283           1
1           97.234714           1
2          112.953771           1
3          130.460597           1
4           95.316933           1
..                ...         ...
95          70.729701           1
96         105.922406           1
97         105.221105           1
98         100.102269           1
99          95.308257           1

[98 rows x 2 columns]
Previsão para os próximos 3 dias: [37.12727563 36.9992055  36.87113536]
Usuário: O que é um investimento?
Chatbot: Investimento é quando você usa dinheiro para comprar algo esperando que ele valha mais no futuro.


Aqui está a explicação dos resultados gerados pelo código:

### **1. Detecção de Fraudes e Transações Válidas**
A função `detect_transactions()` utilizou o **Isolation Forest**, um modelo de machine learning baseado em anomalias, para classificar transações como fraudulentas ou válidas.

#### **Transações Fraudulentas Detectadas**
Essas transações foram classificadas com **fraudulent = -1**, indicando que o modelo identificou padrões suspeitos. Os valores **500, 600, 750 e 1000** foram claramente considerados fraudes porque são **muito maiores do que a média das outras transações**, o que sugere que são outliers. Além disso, duas outras transações, **137.04 e 47.60**, também foram consideradas suspeitas, possivelmente por estarem em extremos da distribuição.

#### **Transações Válidas Detectadas**
As transações restantes foram classificadas com **fraudulent = 1**, indicando que foram consideradas normais pelo modelo. Esses valores estão mais próximos da média do conjunto de dados (em torno de 100, com desvio padrão de 20), o que justifica sua classificação como legítimos.

---

### **2. Previsão de Tendência de Mercado**
A função `market_trend_prediction()` aplicou um modelo de **Regressão Linear** para prever os preços de mercado nos próximos três dias. O modelo indicou que os preços esperados são:

- **Dia 101:** 37.13
- **Dia 102:** 37.00
- **Dia 103:** 36.87

Esses valores sugerem uma **ligeira tendência de queda**, baseada nos dados históricos de preços simulados.

---

### **3. Chatbot Financeiro**
O chatbot respondeu à pergunta **"O que é um investimento?"** com a definição simples:

> "Investimento é quando você usa dinheiro para comprar algo esperando que ele valha mais no futuro."

Isso exemplifica como a IA pode ser usada para automatizar consultas financeiras comuns.